### PR TITLE
[libc++] Fix vector<const T>

### PR DESCRIPTION
**Pull request by philnik777**
#80558 introduced code that assumed that the element type of `vector` is never const. This fixes it and adds a test. Eventually we should remove the `allocator<const T>` extension.

### CODE REVIEW

**Review by ldionne — APPROVED** [review]
LGTM. I think we have an issue tracking that removal, and I might even have a patch locally.

**Comment by pranavk** [comment]
Can we merge this? The original commit has broken lot of our internal workloads and we would like to get this in to alleviate some pain.

**Comment by pranavk** [comment]
Apologies. 

I asked but then thought you may be off for the day.

I did check the CI before merging though. Everything was green except one workload which was stuck on "Waiting for the agent" for "FreeBSD 13" for last 7 hours. Since everything else was green, it seemed reasonable to go ahead with this one.

**Comment by ldionne** [comment]
@philnik777 It’s actually worrying that so many people have commit access to the monorepo. I just checked and about 1800 people have access — arguably just a small fraction of those should really need access now that we have github PRs.

**Comment by rnk** [comment]
To provide a different perspective, I am extremely happy to learn today that this landed yesterday, and we haven't lost a day's worth of test results buried under `std::vector<const T>` issues. It seems entirely reasonable to bypass standard procedure to recover from breakage, and then to reflect afterwards on what could have gone better, and to always assume positive intent.

**Comment by philnik777** [comment]
@rnk I don't think anyone here tried to imply none-positive intent here. I actually watched the CI to land it as soon as the CI is ready. It would have been at most another hour or two until it landed. To me it mostly seemed very weird to first ask whether the patch could be landed and then landing the patch a minute later. I'm not necessarily against landing a patch sooner if it would fix wide-spread breakage (I'm not sure whether this did - I only received a single report), but in this case it wouldn't have been that much longer until the CI was happy and I would have landed it myself. I also would have checked some "land as soon as the CI is green" tick mark if it existed. And FWIW I'm available most of the time during US office hours.

**Comment by ldionne** [comment]
> To provide a different perspective, I am extremely happy to learn today that this landed yesterday, and we haven't lost a day's worth of test results buried under `std::vector<const T>` issues. It seems entirely reasonable to bypass standard procedure to recover from breakage, and then to reflect afterwards on what could have gone better, and to always assume positive intent.

I wasn’t assuming bad intent. I just think it’s surprising that someone who we’ve never (to my recollection ) interacted with in the context of libc++ would have sufficient permissions to merge code into libc++ just like any regular maintainer, and bypass CI checks. I don’t think my surprise is an overreaction, I would argue that LLVM is an outlier in that respect.

In this case there’s no actual problem, the FreeBSD CI was fine to ignore and this simply returned things to green faster. My observation is a general one about how liberal we are with permissions and I do think that’s something we can improve on to make the project run smoother (even outside of security aspects).

**Comment by pranavk** [comment]
@philnik777 Thanks for explaining. I will just add more details to my previous comment in case I wasn't clear. I understand it may seem little weird to first ask and then merge but we were unaware of your working hours and if you were actively monitoring this. I commented on the PR to ask, then looked at your email ID (berlin.de), assumed you were in Berlin time (it was past midnight in Berlin at that time) and off for the day, and went ahead merging it after noticing this PR was created and approved 7+ hours ago at that time.

The original commit had quite a huge impact on us, and we couldn't afford waiting until normal Berlin office hours. Also, CI being mostly green and our internal testing on similar platforms gave me enough confidence to go ahead with this one. I hope that explains my rationale.

**Pull request merged**
